### PR TITLE
Морозов Егор. Задача 1. Вариант 26. Технология SEQ. Линейная фильтрация изображений (блочное разбиение). Ядро Гаусса 3x3.

### PR DESCRIPTION
**Pull request by Egor1dzeN-learn**
Линейная фильтрация изображений — это процесс применения к изображению линейного фильтра (например, свертки с ядром) для выполнения операций, таких как размытие, повышение резкости или выделение границ. В данном случае мы рассмотрим **блочное разбиение изображения** и применение **ядра Гаусса 3x3** для размытия изображения.

---
image

Пример:
```
[[10. 20. 30. 40. 50.]
 [15. 25. 35. 45. 55.]
 [20. 30. 40. 50. 60.]
 [25. 35. 45. 55. 65.]
 [30. 40. 50. 60. 70.]]
```

Размытое изображение:
```
[[13.125 21.25  31.25  41.25  48.125]
 [17.5   26.25  36.25  46.25  53.75 ]
 [22.5   31.25  41.25  51.25  58.75 ]
 [27.5   36.25  46.25  56.25  63.75 ]
 [31.875 41.25  51.25  61.25  68.125]]
```

---

### CODE REVIEW

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+  // Create Task
+  morozov_e_lineare_image_filtering_block_gaussian::TestTaskSequential test_task_sequential(task_data_seq);
+  ASSERT_EQ(test_task_sequential.Validation(), true);
+  test_task_sequential.PreProcessing();
+  test_task_sequential.Run();
+  test_task_sequential.PostProcessing();
+  std::vector<std::vector<double>> real_res = {{1, 6, 7}, {8, 3.875, 1}, {8, 2, 4}};
+  EXPECT_EQ(real_res, image_res);
+}
```
Comment (MoiseevArt) [inline tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp:252]: add tests with negative values.

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+bool morozov_e_lineare_image_filtering_block_gaussian::TestTaskSequential::RunImpl() {
+  // Ядро Гаусса 3x3
+  const std::vector<std::vector<double>> kernel = {
+      {1.0 / 16, 2.0 / 16, 1.0 / 16}, {2.0 / 16, 4.0 / 16, 2.0 / 16}, {1.0 / 16, 2.0 / 16, 1.0 / 16}};
+  for (int i = 0; i < n_; ++i) {
+    for (int j = 0; j < m_; ++j) {
+      if (i == 0 || j == 0 || i == n_ - 1 || j == m_ - 1) {
+        res_[i][j] = input_[i][j];
```
Comment (22380240) [inline tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp:36]:
```suggestion
        res_[i * width + j] = input_[i * width + j];
```

**Review by 22380240 — CHANGES_REQUESTED** [review]

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  task_data_seq->outputs_count.emplace_back(m);
+
+  // Create Task
+  morozov_e_lineare_image_filtering_block_gaussian::TestTaskSequential test_task_sequential(task_data_seq);
+  ASSERT_EQ(test_task_sequential.Validation(), false);
+}
+TEST(morozov_e_lineare_image_filtering_block_gaussian, main_test1) {
+  int n = 5;
+  int m = 5;
+  std::vector<double> image = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
```
Comment (vitaliyvoroshilov) [inline tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp:70]: Its hard to read in this form, you can write like this:
```suggestion
  int n = 5;
  int m = 5;
  // clang-format off
  std::vector<double> image = 
  {1, 1, 1, 1, 1,
   1, 1, 1, 1, 1,
   1, 1, 1, 1, 1,
   1, 1, 1, 1, 1,
   1, 1, 1, 1, 1};
   // clang-format on
```

**Review by vitaliyvoroshilov — CHANGES_REQUESTED** [review]

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp (hunk 3 above)
Comment (Egor1dzeN-learn) [reply tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp]: fixed

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp (hunk 2 above)
Comment (Egor1dzeN-learn) [reply tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp]: fixed

**Review by 22380240 — APPROVED** [review]

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      {1.0 / 16, 2.0 / 16, 1.0 / 16},
+      {2.0 / 16, 4.0 / 16, 2.0 / 16},
+      {1.0 / 16, 2.0 / 16, 1.0 / 16}};
+  // clang-format on
+  for (int i = 0; i < n_; ++i) {
+    for (int j = 0; j < m_; ++j) {
+      if (i == 0 || j == 0 || i == n_ - 1 || j == m_ - 1) {
+        res_[(i * m_) + j] = input_[(i * m_) + j];
+      } else {
+        // std::cout<<i<<''<<j<<"\n";
```
Comment (aobolensk) [inline tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp:36]:
```suggestion
```

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,274 @@
+#include <gtest/gtest.h>
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/morozov_e_lineare_image_filtering_block_gaussian/include/ops_seq.hpp"
+
```
Comment (aobolensk) [inline tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp:9]: Please, random tests here as well

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/perf_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+#include "seq/morozov_e_lineare_image_filtering_block_gaussian/include/ops_seq.hpp"
+
+TEST(morozov_e_lineare_image_filtering_block_gaussian_seq, test_pipeline_run) {
+  int n = 4000;
+  int m = 4000;
+  std::vector<double> image(n * m, 1.0);
+  std::vector<double> image_res(n * m);
+  for (int i = 0; i < n; ++i) {
+    for (int j = 0; j < m; j++) {
+      image_res[(i * m) + j] = 0.0;
```
Comment (aobolensk) [inline tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/perf_tests/main.cpp:19]: Please, use random values there

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/perf_tests/main.cpp (hunk 6 above)
Comment (Egor1dzeN-learn) [reply tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/perf_tests/main.cpp]: fixed

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp (hunk 5 above)
Comment (Egor1dzeN-learn) [reply tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp:27]: fixed

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp (hunk 4 above)
Comment (Egor1dzeN-learn) [reply tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/src/ops_seq.cpp]: fixed

**Review by aobolensk — APPROVED** [review]

File: tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+#include <random>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+#include "seq/morozov_e_lineare_image_filtering_block_gaussian/include/ops_seq.hpp"
+namespace {
+std::vector<double> GenerateRandomVector(int n, int m) {
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> distrib(0, 100);
```
Comment (allnes) [inline tasks/seq/morozov_e_lineare_image_filtering_block_gaussian/func_tests/main.cpp:14]: image range is [0,255]

**Review by allnes — APPROVED** [review]
